### ISSUE TITLE
Setting receptive field size for music training samples

### ISSUE THREAD

**Issue by kaihuchen** [ibab/tensorflow-wavenet#161]
I am trying to apply WaveNet to music samples, and per the WaveNet paper I should be training with a larger receptive field. I feel that ideally the receptive field size should be in the tens of seconds so that WaveNet can keep it coherent across a longer time frame. To keep it simple let's say I just want the receptive field to cover around 5 seconds for now, which translates into 80000 samples in the fields, which then translates into around 16 dilation layers (80000 ~= 2**16.288), which appears to be prohibitively expensive to run. 

Does anybody have experience try using a fairly large receptive field size on music samples? Do you have anything to share?

**Comment by jyegerlehner** [comment]
I think extending the receptive field to anywhere near that large will require a context stack.

**Comment by kaihuchen** [comment]
@jyegerlehner Good point. The idea of 'context stacks' was mentioned only briefly in the paper in terse terms, and my understanding is that it roughly amounts to using multiple smaller WaveNet stacks to handle components of the training data at different frequencies, possibly including polling. 
I am leery of this approach because: while the beauty of WaveNet is that it works directly on raw audios, there is only so far that you can push it. Pooling works well with images, but with audios (and without converting data to the frequency domain) it will destroy vital information in the data. Using much lower sampling rate also grossly distorts the training data. On the other hand, if  training data is converted to the frequency domain first (say, using Fourier Transform) then fed into a variation of the WaveNet, then applying pooling and/or lower sampling rate (in the frequency domain) could actually work much better.
Would love to be proven wrong with the above. So far I have not been able to find any detailed discussion over the Internet regarding the viability of the "context stacks" idea as mentioned in the paper.

**Comment by jyegerlehner** [comment]
I think the pooling bit is optional. And it would only be performed on the longer-range wavenet. 

Well, if you don't want to try context stacks, another alternative to just adding more dilation stacks, is to have each dilation stack grow to a larger size. Instead of:
`1 2 4 ... 256 512 1 2 4 ...'`

in a stack, you could do:

`1 2 4 ... 256 512 1024 2048 4096 1 2 4 ...`

It's shallower than just piling up more dilation stacks that only go up to 512. Might require increasing the number of channels. No idea  how doing that compares with a context stack. There are plenty of hyper parameters to make comparison difficult. And then there's the fact we haven't written the code to do context stacks yet, at least in this project.

> So far I have not been able to find any detailed discussion over the Internet regarding the viability of the "context stacks" idea as mentioned in the paper.

I expect the only ones who have tried it yet are DeepMind.

**Comment by dunnevan** [comment]
One way to increase your effective receptive field without increasing your model size is to decrease the sample rate.  For example 8 KHz is often used in telephone systems, since most vocalizations are bunched at the low end (see G.711). 

Likewise, piano keys peak around 4 KHz, which means 8 KHz is a fair choice according to Nyquist's theorem.

Does anyone have insight into the paper's 16KHz choice?

**Comment by kaihuchen** [comment]
@dunnevan  Good point. Indeed 8KHz should be sufficient for my purpose (i.e., using it as a basis for discovering naive musicality). I have down-sampled some piano music from 44.1K to 8K and it sounds reasonable (though kind of dry due to the lack of higher frequency harmonics, but it will do). Now we shall see how it turns out when cranked through WaveNet.

As to the paper's 16KHz choice, my guess is that 16KHz used to be treated as non-audiophile's hifi-ish sample rate (i.e., good enough for most people), so the paper just pick it up as good enough for its setup too. Most people can tell 8KHz audio as telephone-ish type of lower quality stuff.

**Comment by lemonzi** [comment]
For some reason, GitHub doesn't like me replying from email.... Most of my points are now irrelevant since I see we're all in the same page.

A new contribution: my bet is they chose 16 kHz because they had it already set up for speech. 16 kHz is the standard for wide-band voice applications (most current telephones and VoIP), but has never been widely used in music. Below the 44.1 kHz standard you can find 32 kHz, maybe 22.05 kHz, for low-quality and very compressed music, but I haven't seen it go as low as 16 for applications other than analysis.

Now, what I meant to send a few hours ago:

Well, 8 kHz is enough for speech and used to be common in telephony, but it doesn't sound very clear, especially in the "s" sounds. With 16 kHz you can get decent quality for speech. Using 8 kHz would still be a good idea to speed up the model and see how it performs.

For music it's too low. The highest key of a piano has a fundamental frequency of 4 kHz, but this "frequency" is completely unrelated to the frequency in Nyquist. The frequency in Nyquist includes partials, noise and transients, which go up to 10 kHz easily (the transients go higher). But again, if we are just testing whether the net can figure out "notes" and "pitches" it's a completely viable option.

**Comment by paperkettle** [comment]
I think it'd be extremely useful if the group found a way to 44.1k with a long receptive field possible, while magically keeping it viable for small memory GPUs (aka top of the line consumer gaming Titan XP etc). Even if it means obscene training times. 

I've done a number of experiments training it on full mixes of music at 44.1k, but with the settings commonly used in the repo for 16k. That means my receptive field winds up being really small, and the output just sounds like large grain granular synthesis / timestretching. Which is not especially exciting (to me anyway).

**Comment by kaihuchen** [comment]
Combining input from @jyegerlehner and @dunnevan  I am now testing with the following model:

{
    "filter_width": 2,
    "sample_rate": 8000,
    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536],
    "residual_channels": 32,
    "dilation_channels": 32,
    "quantization_channels": 256,
    "skip_channels": 512,
    "use_biases": true,
    "scalar_input": false,
    "initial_filter_width": 32
}

This is with the aim of having a rather longish receptive field size (~8 seconds?), so that the generated samples from a trained model will exhibit some level of coherence  across a good stretch of time interval.  

I am testing this on an 8 vCPU instance on Google Cloud, which at this time would appear that it will take many days to converge. Stay tuned.

@paperkettle @lemonzi   FYI, I am okay with 8KHz sample rate  because I am simply trying to have WaveNet learn the most basic musical elements, such as the simplest form of note progression, rhythm, and perhaps harmony.  In its current form I don't believe that WaveNet can do much more than that (if at all), but I have hope that a more advanced version of WaveNet could be the core component of a much more sophisticated music composition system one day.

**Comment by Zeta36** [comment]
@kaihuchen, if you are interested in taking into account a receptive field as long as tens os seconds so you can compose congruent music pieces, you will probably need to make use of some attentional interfaces [http://distill.pub/2016/augmented-rnns/#attentional-interfaces](http://distill.pub/2016/augmented-rnns/#attentional-interfaces).

You probably will need to use WaveNet with the higher receptive field possible, but modify also the model so you can take into account (focus) on very old segments of the current song. I don't thing WaveNet can be extended (just setting up its parameters) enough to be able to generate a full congruent song. You will probably need to modify the model greatly if you want to get something really better than the piano sound that DeepMind already show in here: [https://deepmind.com/blog/wavenet-generative-model-raw-audio/](https://deepmind.com/blog/wavenet-generative-model-raw-audio/).

Regards,
Samu.

**Comment by jyegerlehner** [comment]
@kaihuchen 
With that dilation stack, you'll need to set --sample_size. It defaults to 100000, so you won't fill that whole field. You could either set sample_size to the receptive field size, or I think 0 to turn off chunking.

**Comment by kaihuchen** [comment]
@Zeta36 I completely agree with your assessment. I have no illusion that WaveNet in its current form will be able to produce full congruent music. I look at WaveNet's generative capability merely as a way to create useful low-level 'audio features' (conceptually somewhat similar to how CNN uses unsupervised learning to generate useful low-level visual features), and the purpose of my tests here is only to find out how useful such audio features are going to be. There are a lot more work that need to be done in order to produce full congruent music, possibly including the attention interfaces stuff that you mentioned.

**Comment by lemonzi** [comment]
Well, 8 kHz is enough to decipher what someone is saying and used to be
common in telephony, but try it yourself and you'll see it doesn't sound
very clear. With 16 kHz you can get decent quality for speech. Using 8 kHz
would still be a good idea to speed up the model and see how it performs,
only don't expect very high quality at the output.

For music it's too low. The highest key of a piano has a *fundamental*
frequency of 4 kHz, but this "frequency" is completely unrelated to the
frequency in Nyquist. The frequency in Nyquist includes partials, noise and
transients, which IIRC in a piano go up to 10 kHz easily (the transients go
higher); 16 kHz is a compromise for speed and works well enough when there
is only piano present, 22 kHz would be enough for low fidelity. But again,
if we are just testing whether the net can figure out "notes" and "pitches"
it's a viable option.

El dl., 24 oct. 2016 a les 15:03, dunnevan (<notifications@github.com>) va
escriure:

**Comment by LeavesBreathe** [comment]
Just curious, what would happen if you increased the filter_width to 3? You may be able to pick up more features with less layers and therefore, less cost.

**Comment by materialvision** [comment]
@kaihuchen Just curious but how did your experiment with those settings go?

**Comment by nakosung** [comment]
@materialvision It seems that he used cpu only experiement. (8 vCPU instance)

**Comment by kaihuchen** [comment]
@materialvision Even after days of computation it never did converge. I am gradually reducing the receptive field size, tweaking other parameters,  as well as reverting to use the VCTK dataset to see if I can do any better. Perhaps it's time for me to spend a little effort to get a GPU instance going.

**Comment by LeavesBreathe** [comment]
I've tried with 4 titan x's with large field sizes  (10k), and I've found gating to be crucial for convergence.

**Comment by nakosung** [comment]
@LeavesBreathe Have you implemented model parallelism on multiple gpu devices?

**Comment by LeavesBreathe** [comment]
@nakosung Right now I have probably the worst model parallelism design. I'm taking separate layers and placing them on different gpus so I'll have one  1,2,4....10k block on one gpu and the next block on another. Timewise, this is the least efficient as only one gpu runs at a time. However, it does allow all memory usage, which I was interested in (using a big model).

**Comment by kaihuchen** [comment]
@LeavesBreathe    Could you share some details regarding what kind of gating worked for you? Thanks!

**Comment by LeavesBreathe** [comment]
by gating I simply mean the sigmoidal dilated causal conv that is multiplied by the tanh activated dilation. Was anyone else able to do model parallelism on gpus?

**Comment by strongrex2001** [comment]
I tried to extend the receptive field with dialtion 1, 2, ..., 8192 and has 5 stacks. But the loss is always around 3.5, on my set of 480 pieces of absolute music, even when I trained 130K steps. @kaihuchen Maybe similar results to yours
